### ISSUE TITLE
Failure during SAF zip import in some select cases can delete entire directories

### ISSUE THREAD

**Issue by jonas-atmire** [DSpace/DSpace#9330]
**Describe the bug**
This is a very specific bug that I have seen on 2 separate DSpace instances already, and it's a major issue which can result in entire servers being wiped because of the way that the code is structured.

While running the SAF import from a ZIP file: https://wiki.lyrasis.org/display/DSDOC7x/Importing+and+Exporting+Items+via+Simple+Archive+Format#ImportingandExportingItemsviaSimpleArchiveFormat-AddingItemstoaCollectionfromazipfile

This ended up failing because of an issue with the provided zip file to import, resulting in the following error

```
The script has started
Destination collections:
Owning Collection: Bison
java.io.IOException: Unable to delete directory /home/dspace/tmp-bigdata.
        at [org.apache.commons.io](http://org.apache.commons.io/).FileUtils.deleteDirectory(FileUtils.java:1127)
        at org.dspace.app.itemimport.ItemImport.internalRun(ItemImport.java:234)
```
(Note that that tmp-bigdata is a mounted disk, so it couldn't FULLY delete it, but it can be reproduce with a regular directory as well.)

The code tries to completely delete the source and work directory in the finally call here:
*  https://github.com/DSpace/DSpace/blob/dspace-7.4/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImport.java#L231

If you set the ‘source’ directory (The one being cleared), to the home directory for example -> It ‘kills’ the server, and you can’t access it anymore, since that contains everything from a random text file, to the .ssh directory etc.

There is no reason for the code to completely DELETE the entire directory. (IMO, it's not that import's responsibility to clear out those file, and this should be avoided to clear directories)
Fore example, when preparing imports, and there is a directory with x GBs of different import files, if the FIRST import fails -> It would clear out all of the other ones as well.

**To Reproduce**
* Setup a fresh DSpace with at least one collection and eperson
    * Note that at the time of writing, this only was reproduced on 7.4 (7.0, 7.2, 7.6 didn't break it, but the latest could should be investigated as well)
    * 7.6 ALSO tries to do a deletion, but that never completely triggered in our internal tests -> https://github.com/DSpace/DSpace/blob/dspace-7.6/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImport.java#L240
* Create directory to test with (e.g. ~/Documents/test-import)
* Run ./dspace import -a -e {{eperson-email}} -c {{collection-UUID}} -s ~/Documents/test-import -z testing.zip -m temp-mapfile with the -c argument being an existing collection’s UUID and -e argument an existing EPerson. 
    * This is important because it will otherwise not reach the part of the code responsible for the issue, the rest (e.g. the zip) doesn’t need to exist.

**Expected behavior**

Either any failure, or any success should never delete an entire directory worth of files, even IF that directory only contained a single ZIP file being used for the import 

The deletion of previously imported files or directories should not be handled by the aforementioned script

**Comment by tdonohue** [comment]
@jonas-atmire : Pulling this over to the 7.6.x board and flagging as high priority.  Let me know if you/your team wants to volunteer to create a fix.

**Comment by mwoodiupui** [comment]
The root of the problem IMHO is that we are unpacking Zip archives instead of just reading them.  This leads to a number of security problems (such as destroying part or all of a running DSpace) as well as a requirement for empty storage of more than double the size of a Zip submission packet and the necessity of cleaning up the unpacked files.  The JRE has classes that model Zip archives quite well -- we should use them.

**Comment by jonas-atmire** [comment]
@mwoodiupui I agree that the full unpack and subsequent deletion of those unpacked files is not really the way to go in most cases.
I've had multiple occasions where the following config had to be updated to a specifically mounted storage since the exported zip was just too large to handle in the 'regular' server.
* https://github.com/DSpace/DSpace/blob/main/dspace/config/dspace.cfg#L893-L895

I do wonder how well the existing JRE classes might handle large ZIP files though, as we want to avoid overloading the server and impacting the regular usage. (That IS one thing that the extraction of the ZIP can avoid in some cases, which might be the reason for it being implemented as such in the first place)

Aside from that, even if we would NOT go for the rework of how these ZIPs are used, the 'work dir' is user specific and created on the fly, so the deletion of that would not be an "issue", but this should at least start with NOT clearing the sourcedir

**Comment by tdonohue** [comment]
@jonas-atmire and @mwoodiupui :  I agree that it might be nice to handle this without a full extraction.  But, I'd caution about waiting for a perfect solution.  If someone is available to investigate how the JRE can handle large zips, than that would be great.  But, if no one can step forward at this time to do that investigation, then I'd suggest maybe implementing a "quick fix" of simply not deleting the sourcedir (if that's the main thing that is wrong with the current code).

**Comment by jonas-atmire** [comment]
@tdonohue 
# Intermediate fixes
We've gone through the small intermediate fixes for the for 7.4 AND 7.6 (7.6 did not initially fail, but this is more to be on the safe side since 7.4 also fails on a very specific case, so it's not that we didn't break 7.6 now that it's not broken if that makes sense)

## 7.4 
* Commit: https://github.com/atmire/DSpace/commit/273906419718461d882b722a2bc88bca695c2f26
    * Tests are to be fixed -> They succeed locally, but on github they are failing  
## 7.6 
* Commit: https://github.com/atmire/DSpace/commit/df2382e13bd69ca1071a2f8786d18856fc11310f
# Analysis
With regards to reading the info straight from the ZIP instead of first extracting it, it seems pretty easy to find information about that, e.g. using [ZipFile and ZipEntry](https://www.baeldung.com/java-read-zip-files).

This would need to happen in [ItemImportServiceImpl](https://github.com/DSpace/DSpace/blob/dspace-7.4/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportServiceImpl.java#L240) where several methods like “addItems” accept a sourceDir where the zip was extracted. We would remove the [unzip in ItemImport](https://github.com/DSpace/DSpace/blob/dspace-7.4/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImport.java#L311) and instead pass a ZipFile or something to the service methods.
These methods would then be able to read the ZipFile’s ZipEntries instead of reading the unzipped content.

Since this requires both the code changes, IT testing, manual testing, and updating the docs accordingly, I think this requires at least 3 days of work

**Comment by tdonohue** [comment]
@jonas-atmire : Please just send along a PR when you are ready -- I won't have time to dig into this further for at least a few weeks as we are busy on major dependency updates for 8.0.  

In general, I'm perfectly OK with just a simple fix.  I'm not against the larger refactor if the Atmire team wants to do it. But that larger refactor would be contributed as volunteer work (unpaid/unfunded), so that's a decision for Atmire as to whether you see a purpose to spending your time implementing the larger change.

**Comment by mwoodiupui** [comment]
If nobody else does the larger refactor, I'll probably do it at some point.

**Comment by tdonohue** [comment]
Closing, as noted in [my PR review](https://github.com/DSpace/DSpace/pull/9392#issuecomment-2103472010), I believe this was already fixed by #8870

>@jonas-atmire : I finally got around to reviewing this, and I think it's already been fixed on 7.6 by this PR https://github.com/DSpace/DSpace/pull/8870. I believe that's why you noted this isn't easily reproducible in 7.6 (in the https://github.com/DSpace/DSpace/issues/9330#issue-2134350898).
>
> Correct me if I'm wrong, but I think the core issue you were hitting was described by @nwoodward in this comment: https://github.com/DSpace/DSpace/issues/8759#issuecomment-1500327216 . The summary is that, prior to 7.6, the UI and CLI had different behaviors. But, as of https://github.com/DSpace/DSpace/pull/8870, both should leave around the shared source directory.

If I'm wrong, then we can reopen this. But, I believe this is fixed in 7.6+ and have not found a way to reproduce it.
